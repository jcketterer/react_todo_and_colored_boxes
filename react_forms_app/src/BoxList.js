import React, { useState } from 'react'
import Box from './Box'
import NewBoxForm from './NewBoxForm'
import './BoxList.css'

const BoxList = () => {
  const [boxes, setBoxes] = useState([])
  const add = boxObj => {
    setBoxes(boxes => [...boxes, boxObj])
  }

  const remove = id => {
    setBoxes(boxes => boxes.filter(box => box.id !== id))
  }

  const boxAttributes = boxes.map(box => (
    <Box
      key={box.id}
      id={box.id}
      width={box.width}
      height={box.height}
      handleRemove={remove}
      backgroundColor={box.backgroundColor}
    />
  ))

  return (
    <div className="BoxList">
      <NewBoxForm createBox={add} />
      <div className="BoxList-Box">{boxAttributes}</div>
    </div>
  )
}

export default BoxList
