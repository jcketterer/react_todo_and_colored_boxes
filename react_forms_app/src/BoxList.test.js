import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import BoxList from './BoxList'

const addBox = (height = 2, width = 2, color = 'peachpuff') => {
  const heightInput = screen.getByLabelText('Height:')
  const widthInput = screen.getByLabelText('Width:')
  const backgroundInput = screen.getByLabelText('Background Color:')

  fireEvent.change(backgroundInput, { target: { value: color } })
  fireEvent.change(widthInput, { target: { value: width } })
  fireEvent.change(heightInput, { target: { value: height } })

  const btn = screen.getByText('Add your Box!')

  fireEvent.click(btn)
}

it('should render without breaking', () => {
  render(<BoxList />)
})

it('matches snap', () => {
  const { asFragment } = render(<BoxList />)
  expect(asFragment()).toMatchSnapshot()
})

it('should add a box', () => {
  const boxList = render(<BoxList />)

  expect(boxList.queryByText('Remove your box!')).not.toBeInTheDocument()

  addBox(boxList)

  const remove = boxList.queryByText('Remove your box!')
  expect(remove).toBeInTheDocument()
  expect(boxList.getAllByDisplayValue('')).toHaveLength(3)
})

it('should be able to remove a box', () => {
  const boxList = render(<BoxList />)
  addBox(boxList)

  const remove = boxList.queryByText('Remove your box!')

  fireEvent.click(remove)
  expect(remove).not.toBeInTheDocument
})
