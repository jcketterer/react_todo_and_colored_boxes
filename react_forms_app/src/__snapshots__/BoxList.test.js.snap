// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snap 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <div
      class="NewBoxForm"
    >
      <form>
        <div
          class="NewBoxFormInput"
        >
          <label
            for="height"
          >
            Height: 
          </label>
          <input
            id="height"
            name="height"
            type="text"
            value=""
          />
        </div>
        <div
          class="NewBoxFormInput"
        >
          <label
            for="width"
          >
            Width: 
          </label>
          <input
            id="width"
            name="width"
            type="text"
            value=""
          />
        </div>
        <div
          class="NewBoxFormInput"
        >
          <label
            for="backgroundColor"
          >
            Background Color: 
          </label>
          <input
            id="backgroundColor"
            name="backgroundColor"
            type="text"
            value=""
          />
        </div>
        <button
          id="newBoxButton"
        >
          Add your Box!
        </button>
      </form>
    </div>
    <div
      class="BoxList-Box"
    />
  </div>
</DocumentFragment>
`;
